### PR TITLE
Various fixes for wxQt wxBookCtrl, wxTreeCtrl, and for C++20 MSVC build.

### PR DESCRIPTION
**Pull request by dsa-t**

### CODE REVIEW

File: src/qt/treectrl.cpp [hunk 1]
```diff
@@ -355,7 +355,8 @@ class wxQTreeWidget : public wxQtEventSignalHandler<QTreeWidget, wxTreeCtrl>
         );
 
         changingEvent.SetOldItem(wxQtConvertTreeItem(previous));
-        EmitEvent(changingEvent);
+        if (EmitEvent(changingEvent) && !changingEvent.IsAllowed())
+            return;
```
Comment (AliKet) [inline src/qt/treectrl.cpp:360]: But the current item will not be restored at `Qt` level below if this event is disallowed at `wx `level ?

File: src/qt/treectrl.cpp [hunk 2]
```diff
@@ -374,7 +375,7 @@ class wxQTreeWidget : public wxQtEventSignalHandler<QTreeWidget, wxTreeCtrl>
             wxQtConvertTreeItem(current)
         );
         changedEvent.SetOldItem(wxQtConvertTreeItem(previous));
-        wxPostEvent(treeCtrl, changedEvent);
+        EmitEvent(changedEvent);
     }
```
Comment (AliKet) [inline src/qt/treectrl.cpp:379]: According to the comment above these lines, using `wxPostEvent` is intentional! but if using `EmitEvent`  here doesn't break anything, then the comment should be removed.

File: build/cmake/toolkit.cmake [hunk 3]
```diff
@@ -141,6 +141,8 @@ if(WXQT)
         set(CMAKE_SHARED_LIBRARY_SUFFIX_C ${CMAKE_SHARED_LIBRARY_SUFFIX})
         set(CMAKE_SHARED_LIBRARY_SUFFIX_CXX ${CMAKE_SHARED_LIBRARY_SUFFIX})
     endif()
+
+    list(APPEND wxTOOLKIT_DEFINITIONS __WXQT__)
```
Comment (MaartenBent) [inline build/cmake/toolkit.cmake:145]: This is/should already be set in
https://github.com/wxWidgets/wxWidgets/blob/0e4a5d9a0d748a4e5ae475e3c83fc988076a3098/build/cmake/toolkit.cmake#L57

Comment (dsa-t) [reply build/cmake/toolkit.cmake]: You're right.

File: src/qt/treectrl.cpp (hunk 1 above)
Comment (dsa-t) [reply src/qt/treectrl.cpp:376]: It didn't seem to work even before. But now I've added changes that work.

File: src/qt/treectrl.cpp (hunk 2 above)
Comment (dsa-t) [reply src/qt/treectrl.cpp:392]: Updated with another method that works on my config.

File: src/qt/treectrl.cpp (hunk 1 above)
Comment (AliKet) [reply src/qt/treectrl.cpp:376]: Could you please try this patch and let me know how it goes with you. as there is still an issue that can be seen when a selection is vetoed, which does not happen with the provided patch, TIA.

```c++
diff --git a/src/qt/treectrl.cpp b/src/qt/treectrl.cpp
index 5a9421c2d8..e84c195597 100644
--- a/src/qt/treectrl.cpp
+++ b/src/qt/treectrl.cpp
@@ -125,6 +125,52 @@ private:
 
 Q_DECLARE_METATYPE(TreeItemDataQt)
 
+class wxQTreeWidget;
+
+class wxQItemSelectionModel : public QItemSelectionModel
+{
+public:
+    wxQItemSelectionModel(QAbstractItemModel* model, wxQTreeWidget* parent);
+
+    virtual void
+    select(const QItemSelection &selection, QItemSelectionModel::SelectionFlags command)
+    {
+        if ( ms_isSelectionChangeAllowed )
+        {
+            QItemSelectionModel::select(selection, command);
+        }
+    }
+
+    virtual void
+    select(const QModelIndex &index, QItemSelectionModel::SelectionFlags command) override
+    {
+        if ( ms_isSelectionChangeAllowed )
+        {
+            QItemSelectionModel::select(index, command);
+        }
+    }
+
+    virtual void
+    setCurrentIndex(const QModelIndex& index, QItemSelectionModel::SelectionFlags command) override
+    {
+        ms_isSelectionChangeAllowed = IsSelectionChangeAllowed(index);
+
+        if ( ms_isSelectionChangeAllowed )
+        {
+            QItemSelectionModel::setCurrentIndex(index, command);
+        }
+    }
+
+private:
+    bool IsSelectionChangeAllowed(const QModelIndex& index) const;
+
+    wxQTreeWidget* m_treeWidget;
+    static bool ms_isSelectionChangeAllowed;
+};
+
+/*static*/
+bool wxQItemSelectionModel::ms_isSelectionChangeAllowed = true;
+
 class wxQTreeWidget : public wxQtEventSignalHandler<QTreeWidget, wxTreeCtrl>
 {
 public:
@@ -149,6 +195,15 @@ public:
         viewport()->setAcceptDrops(true);
         setDropIndicatorShown(true);
         setEditTriggers(QAbstractItemView::SelectedClicked);
+
+        auto selection_model = new wxQItemSelectionModel(model(), this);
+        connect(model(), SIGNAL(destroyed()), selection_model, SLOT(deleteLater()));
+        setSelectionModel(selection_model);
+
+        connect(selectionModel(), SIGNAL(currentChanged(QModelIndex, QModelIndex)),
+                this, SLOT(_q_emitCurrentItemChanged(QModelIndex, QModelIndex)));
+        connect(selectionModel(), SIGNAL(selectionChanged(QItemSelection, QItemSelection)),
+                this, SLOT(_q_selectionChanged(QItemSelection, QItemSelection)));
     }
 
     virtual void paintEvent (QPaintEvent * event)
@@ -330,6 +385,9 @@ protected:
         QTreeWidget::closeEditor(editor, hint);
     }
 
+    friend class wxQItemSelectionModel;
+    using QTreeWidget::itemFromIndex;
+
 private:
     void ReplaceIcons(QTreeWidgetItem *item)
     {
@@ -348,33 +406,13 @@ private:
     {
         wxTreeCtrl* treeCtrl = GetHandler();
 
-        wxTreeEvent changingEvent(
-            wxEVT_TREE_SEL_CHANGING,
-            treeCtrl,
-            wxQtConvertTreeItem(current)
-        );
-
-        changingEvent.SetOldItem(wxQtConvertTreeItem(previous));
-        EmitEvent(changingEvent);
-
-        if ( !changingEvent.IsAllowed() )
-        {
-            wxQtEnsureSignalsBlocked blocker(this);
-            setCurrentItem(previous);
-            return;
-        }
-
-        // QT doesn't update the selection until this signal has been
-        // processed. Deferring this event ensures that
-        // wxTreeCtrl::GetSelection returns the new selection in the
-        // wx event handler.
         wxTreeEvent changedEvent(
             wxEVT_TREE_SEL_CHANGED,
             treeCtrl,
             wxQtConvertTreeItem(current)
         );
         changedEvent.SetOldItem(wxQtConvertTreeItem(previous));
-        wxPostEvent(treeCtrl, changedEvent);
+        EmitEvent(changedEvent);
     }
 
     void OnItemActivated(QTreeWidgetItem *item, int WXUNUSED(column))
@@ -553,6 +591,31 @@ private:
     QPixmap m_placeHolderImage;
 };
 
+wxQItemSelectionModel::wxQItemSelectionModel(QAbstractItemModel* model,
+                                             wxQTreeWidget* parent)
+    : QItemSelectionModel(model, parent), m_treeWidget(parent)
+{
+}
+
+bool wxQItemSelectionModel::IsSelectionChangeAllowed(const QModelIndex& index) const
+{
+    if ( !index.isValid() )
+        return true;
+
+    wxTreeCtrl* treeCtrl = m_treeWidget->GetHandler();
+
+    wxTreeEvent changingEvent(
+        wxEVT_TREE_SEL_CHANGING,
+        treeCtrl,
+        wxQtConvertTreeItem(m_treeWidget->itemFromIndex(index))
+    );
+
+    changingEvent.SetOldItem(treeCtrl->GetFocusedItem());
+    m_treeWidget->EmitEvent(changingEvent);
+
+    return changingEvent.IsAllowed();
+}
+
 wxTreeCtrl::wxTreeCtrl() :
     m_qtTreeWidget(nullptr)
 {
```

Comment (dsa-t) [reply src/qt/treectrl.cpp:376]: It seems like `GetSelection` doesn't return the new value after successful `OnSelChanged`. But Veto works.

```
18:06:54: OnSelChanging("Root")
18:06:54: GetSelection: ()
18:06:54: OnSelChanged("Root")
18:06:54: GetSelection ()
```

Using this `treetest` patch:
```patch
diff --git a/samples/treectrl/treetest.cpp b/samples/treectrl/treetest.cpp
index 85439caf2c..4b09cfb27a 100644
--- a/samples/treectrl/treetest.cpp
+++ b/samples/treectrl/treetest.cpp
@@ -1326,11 +1326,13 @@ void MyTreeCtrl::DoResetBrokenStateImages(const wxTreeItemId& idParent,
 void MyTreeCtrl::LogEvent(const wxString& name, const wxTreeEvent& event)
 {
     wxTreeItemId item = event.GetItem();
+    wxTreeItemId olditem = event.GetOldItem();
     wxString text;
     if ( item.IsOk() )
         text << '"' << GetItemText(item).c_str() << '"';
-    else
-        text = "invalid item";
+    if (olditem.IsOk())
+        text << " old " << '"' << GetItemText(olditem).c_str() << '"';
+
     wxLogMessage("%s(%s)", name, text);
 }
 
@@ -1349,11 +1351,40 @@ TREE_EVENT_HANDLER(OnSetInfo)
 TREE_EVENT_HANDLER(OnItemExpanded)
 TREE_EVENT_HANDLER(OnItemExpanding)
 TREE_EVENT_HANDLER(OnItemCollapsed)
-TREE_EVENT_HANDLER(OnSelChanged)
-TREE_EVENT_HANDLER(OnSelChanging)
+//TREE_EVENT_HANDLER(OnSelChanged)
+//TREE_EVENT_HANDLER(OnSelChanging)
 
 #undef TREE_EVENT_HANDLER
 
+void MyTreeCtrl::OnSelChanging( wxTreeEvent& event)
+{
+    LogEvent("OnSelChanging", event);
+
+    wxTreeItemId item = GetSelection();
+    wxString text;
+    if (item.IsOk())
+        text << '"' << GetItemText(item).c_str() << '"';
+    
+    wxLogMessage("%s(%s)", "GetSelection: ", text);
+
+    if (GetItemText(event.GetItem()).EndsWith("3"))
+        event.Veto();
+}
+
+void MyTreeCtrl::OnSelChanged(wxTreeEvent& event)
+{
+    LogEvent("OnSelChanged", event);
+
+
+    wxTreeItemId item = GetSelection();
+    wxString text;
+    if (item.IsOk())
+        text << '"' << GetItemText(item).c_str() << '"';
+
+    wxLogMessage("%s(%s)", "GetSelection: ", text);
+}
+
+
 void LogKeyEvent(const wxString& name, const wxKeyEvent& event)
 {
     wxString key;

```

Comment (AliKet) [reply src/qt/treectrl.cpp:376]: Should be fixed by changing `GetSelection()` implementation to return the focued item as done under **wxMSW**:
```c++
wxTreeItemId wxTreeCtrl::GetSelection() const
{
    wxCHECK_MSG( !HasFlag(wxTR_MULTIPLE), wxTreeItemId(),
                 "this only works with single selection controls" );

    return GetFocusedItem();
}
``` 
Thanks for testing the patch by the way.

Comment (dsa-t) [reply src/qt/treectrl.cpp:376]: Yes, this works for `GetSelection`, but `GetSelections` is still lagging behind.

Test patch: [treetest2.patch](https://github.com/user-attachments/files/15686750/treetest2.patch)

Comment (AliKet) [reply src/qt/treectrl.cpp:376]: Indeed, the selections returned by `GetSelections()` are actually updated after the execution of the `currentItemChanged()` signal where the `wxEVT_TREE_SEL_CHANGED` event is generated (i.e. when the `itemSelectionChanged( )` is emitted), which is too late for the selections to be up to date.

I'll try to find a solution...

Comment (AliKet) [reply src/qt/treectrl.cpp:376]: I've updated the patch above in https://github.com/wxWidgets/wxWidgets/pull/24586#discussion_r1628382776 and I think everything is working now as expected. but let me know if this is still broken.
Thanks a lot for your testing and contribution.

File: src/qt/listctrl.cpp [hunk 4]
```diff
@@ -1836,14 +1836,33 @@ void wxListCtrl::CheckItem(long item, bool check)
...
-void wxListCtrl::SetSingleStyle(long WXUNUSED(style), bool WXUNUSED(add))
+void wxListCtrl::SetSingleStyle(long style, bool add)
 {
+    long flag = GetWindowStyleFlag();
+
+    // Get rid of conflicting styles
+    if (add)
+    {
+        if (style & wxLC_MASK_TYPE)
+            flag = flag & ~wxLC_MASK_TYPE;
```
Comment (vadz) [inline src/qt/listctrl.cpp:1847]: Very minor, but personally I find
```suggestion
            flag &= ~wxLC_MASK_TYPE;
```
more readable.

File: src/qt/listctrl.cpp [hunk 5]
```diff
@@ -1836,14 +1836,33 @@ void wxListCtrl::CheckItem(long item, bool check)
...
+        flag &= ~style;
+
+    SetWindowStyleFlag(flag);
 }
 
 void wxListCtrl::SetWindowStyleFlag(long style)
 {
     m_windowStyle = style;
-    m_qtTreeWidget->setHeaderHidden((style & wxLC_NO_HEADER) != 0);
+    m_qtTreeWidget->setHeaderHidden((style & (wxLC_NO_HEADER | wxLC_LIST)) != 0);
```
Comment (vadz) [inline src/qt/listctrl.cpp:1865]: Why only `wxLC_LIST`? We should also hide it for `wxLC_ICON` and `wxLC_SMALL_ICON`, so I'd do something like

```suggestion
    m_qtTreeWidget->setHeaderHidden((style & wxLC_NO_HEADER) != 0 || (style & wxLC_REPORT) == 0);
```

File: src/qt/treectrl.cpp [hunk 6]
```diff
@@ -348,33 +349,46 @@ class wxQTreeWidget : public wxQtEventSignalHandler<QTreeWidget, wxTreeCtrl>
...
         changingEvent.SetOldItem(wxQtConvertTreeItem(previous));
-        EmitEvent(changingEvent);
 
-        if ( !changingEvent.IsAllowed() )
+        if (EmitEvent(changingEvent) && !changingEvent.IsAllowed())
         {
-            wxQtEnsureSignalsBlocked blocker(this);
-            setCurrentItem(previous);
+            // Has to be deferred, otherwise it doesn't have any effect.
+            QTimer::singleShot(0,
```
Comment (vadz) [inline src/qt/treectrl.cpp:368]: Is this better than using `CallAfter()`?

**Review by vadz — COMMENTED** [review]
I didn't follow the discussion about the selected item change (or not), I wonder if we have any unit tests covering this or if one could be added easily by chance, as whatever the answer is, i.e. whether it works or not now, this definitely doesn't look "obviously correct".

All the other changes do, thanks for making them!

File: src/qt/treectrl.cpp (hunk 1 above)
Comment (dsa-t) [reply src/qt/treectrl.cpp]: Thanks, it is working!
I've dropped my commit (https://github.com/dsa-t/wxWidgets/commit/7a7433ff4f7627c5fc298a4a605d1b1f12511a54) from this PR.
You can create your own PR now.

File: src/qt/treectrl.cpp (hunk 6 above)
Comment (dsa-t) [reply src/qt/treectrl.cpp]: This is no longer needed with AliKet's patch.

File: src/qt/listctrl.cpp (hunk 4 above)
Comment (dsa-t) [reply src/qt/listctrl.cpp:1847]: It was copied from wxMSW implementation basically.

Comment (dsa-t) [reply src/qt/listctrl.cpp]: Updated.

**Pull request merged**